### PR TITLE
Make builders for Stream and View lazier

### PR DESCRIPTION
**Pull request by NthPortal**
follow-up to #9466

cc @scala/collections

### CODE REVIEW

**Review by dwijnand — COMMENTED** [review]
A soft LGTM

**Comment by NthPortal** [comment]
oh, it seems I've broken the ability to call `result()` on a `Stream` builder multiple times. I'll see if there's an easy fix or workaround

**Comment by johnynek** [comment]
looks like there is a test that this breaks:

https://travis-ci.com/github/scala/scala/jobs/480191232#L1851

https://github.com/scala/scala/blob/ccb7657760f72d8a48daf2efb1d3db0173425cf7/test/scalacheck/scala/collection/immutable/SeqProperties.scala#L23

File: src/library/scala/collection/immutable/Stream.scala [hunk 1]
```diff
@@ -471,7 +471,7 @@ object Stream extends SeqFactory[Stream] {
 
   def empty[A]: Stream[A] = Empty
 
-  override def newBuilder[A]: mutable.Builder[A, Stream[A]] = ArrayBuffer.newBuilder[A].mapResult(array => from(array))
+  override def newBuilder[A]: mutable.Builder[A, Stream[A]] = Iterator.newBuilder[A].mapResult(fromIterator)
```
Comment (johnynek) [inline src/library/scala/collection/immutable/Stream.scala]: this could use LazyList couldn't it? That would align Stream and LazyList more, which I think would be good.

Comment (NthPortal) [reply src/library/scala/collection/immutable/Stream.scala]: no reason to. `Stream` does its own memoizing, so it's okay that the built iterator can only be used once. `View`, however, needs a backing collection that will memoize the elements.

**Comment by NthPortal** [comment]
> looks like there is a test that this breaks

that's what my comment was about

File: src/library/scala/collection/immutable/Stream.scala (hunk 1 above)
Comment (NthPortal) [reply src/library/scala/collection/immutable/Stream.scala]: and `LazyList`'s builder does not allow you to call `result()` multiple times in a row, so switching to it would not fix the failing test

**Comment by joshlemer** [comment]
I have the same concerns as @julienrf (expressed in https://github.com/scala/scala/pull/7573) that having mutable builders with lazy methods is confusing and inconsistent. The `Builder#addAll` method (which is actually `Growable#addAll`) is part of an interface and a package (`collection.mutable`) that is designed for use in heavily imperative, (at least locally) side-effectful, looping style code. This style does not mix well with lazy transformations, or at least when you do want to use laziness in these situations it should be dead obvious to everyone. 

It may not sound so dangerous when you consider only code that deals directly with concrete builders like `Iterator.newBuilder[A]`, but consider code that deals with builders in general. Normally one who writes against the `Builder` interface will not be thinking about how any of the methods may actually be completely lazy, and it will result in incorrect code if, say, they reuse the underlying collection of the iterator/view, which is a very common practice in imperative code
```scala
def writeTo[C](b: Builder[(String, Int), C]): Unit = {
  val buffer: ArrayBuffer[A] = new ArrayBuffer(64)
  
  while(???) {
    buffer ++= ??? // fill up the buffer while doing some work...
    
    // after the work, maybe transform each element and add it to the builder
    // here we leverage laziness in order to not waste space materializing strict intermediate collections
    b.addAll(buffer.iterator.map(_.toString).zipWithIndex)
    buffer.clear()
  }
}
```
the above code seems reasonable to me, but will get a concurrent modification exception (or perhaps undefined behaviour in the case of using Views instead of Iterators. 

I think that weakening the contract of `Growable/Builder#addAll` reduces its overall utility because callers have no idea how that builder behaves, unless they know exactly what kind of builder they're using, a requirement that undermines the benefits of abstracting out an interface in the first place. 

Also consider the burden this places on custom collection authors. For anyone looking to author a custom lazy collection, they'll have to make sure that they override the default strict implementation of `addAll` with a lazy one in order to follow the precedent being set. To me, that harkens back to the issues with the architecture of the 2.8-2.12 collections where lazy things have to make sure to override every single inherited method because by default everything is strict and so one forgotten override breaks your collection. 

This mixing of lazy and strict (as well as deduplicating/non-deduplicating, but I digress) implementations behind the same interface is also what causes `Iterable` to be used less than it might be (one cannot safely call `Iterable#map` etc, because it is not known if it is lazy or strict, so in practice one should always be transforming to a view/iterator or ensuring it is strict, depending on the caller's specific use-case). That ship has long since sailed of course, but I think it would be a mistake to repeat that decision in the case of Growables/Builders. 

There is a very fair concern about stack safety in the case of View (Iterator and LazyList appear stack-safe for concatenation to me), and I think that that can and should be dealt with directly rather than switching out to builder's for long chains of `concat`.

In short, to me, we have LazyList, Iterator and View operations themselves if we want to do things like lazy concatenation, that's the exact use-case they were designed for! Don't get me wrong either, I <3 laziness when it is appropriate (I've argued that all remaining strict operations on Views become lazy like groupBy, sorted, etc), but Builders/Growables are mutable imperative side-effectful data containers in the `mutable` package where everything is always strict. Let them be strict I say.

**Comment by SethTisue** [comment]
I had been thinking about commenting along similar lines as what Josh has written.

**Comment by SethTisue** [comment]
Josh also expressed similar reservations at https://github.com/scala/scala/pull/9466#discussion_r568040304, so depending on what we decide here, #9466 may need to be reverted.

**Comment by NthPortal** [comment]
my primary issue with the builders being strict is that it makes them... useless. a strict `LazyList`, or a strict `Iterator`, are collections with completely different properties

**Comment by SethTisue** [comment]
> my primary issue with the builders being strict is that it makes them... useless

Could they be removed, then, or must they be present to satisfy some completeness constraint?

> a strict LazyList, or a strict Iterator, are collections with completely different properties

I catch your drift, but in itself that isn't a complete argument.

It's normal to produce an `Iterator` from a strict collection, for example `List(1, 2, 3).iterator`. There's no reason to forbid this. So it isn't obvious to me — unless further justification is provided — that it's wrong to allow the same thing via a builder.

`Iterator.apply` is strict, too. That's another precedent for a result being of a nominally lazy type even though the inputs were strictly evaluated.

**Comment by japgolly** [comment]
I've come around to agreeing with @joshlemer's point above. If one wants to lazily create a View or Stream, those classes already have lazy operators for doing so without the need for a builder.

I've tried to think of a situation where you'd really want that builder to be lazy (and not using a builder would be unacceptable) and the best I can come up with is a situation where some API creates a collection for you using a builder (and the API is such that it doesn't just simply return the built result but rather keeps it internally, maybe even parameterising itself on the collection type). It's an edge-case for sure but a legitimate one.

Seeing as there are legitimate cases for having both strict *and* lazy builders, how about we keep the existing builders strict and add new lazy ones that users access explicitly (eg `View.newLazyBuilder`)?

**Comment by Ichoran** [comment]
I'm convinced by @joshlemer's argument.  As a supporting observation, remember when `Seq` meant `collection.Seq`?  After unpleasant experiences with `collection.Seq` admitting behavior that people didn't intuitively expect, we changed `Seq` to be `collection.immutable.Seq`.  Lazy builders would, I think, be going in the opposite direction.

**Comment by NthPortal** [comment]
> Could they be removed

no, they're part of the contract/api of `IterableFactory`. we could however deprecate the... leaf nodes? `Iterator`, `LazyList`, etc.

**Comment by NthPortal** [comment]
> `Iterator.apply` is strict, too.

not really a fan of that either, tbh, but like `LazyList.apply`, there's no getting around it

**Comment by NthPortal** [comment]
closing this issue, as I largely agree with the consensus. however, we'll have to open a PR to make `LazyList`'s builder strict again (should probably be backed by a `List` builder)

**Comment by julienrf** [comment]
> > `Iterator.apply` is strict, too.
> 
> not really a fan of that either, tbh, but like `LazyList.apply`, there's no getting around it

We have `unfold` methods on those factories.
